django/db/models/functions/date.py

```diff
@@ -0,0 +1,216 @@
+                if self.tzinfo is None:
+                    tzname = timezone.get_current_timezone_name()
+                else:
+                    tzname = timezone._get_timezone_name(self.tzinfo)
+            sql, params = connection.ops.datetime_trunc_sql(self.kind, inner_sql, tzname)
+        elif isinstance(self.output_field, DateField):
+            sql = connection.ops.date_trunc_sql(self.kind, inner_sql)
+            params = []
+        else:
+            raise ValueError('Trunc only valid on Date or DateTime fields')
+        return sql, inner_params + params
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        copy = super(TruncBase, self).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        field = copy.lhs.output_field
+        # DateTimeField is a subclass of DateField so this works for both.
+        assert isinstance(field, DateField), (
+            "%r isn't a DateField or DateTimeField." % field.name
+        )
+        return copy
+
+    def convert_value(self, value, expression, connection, context):
+        if isinstance(self.output_field, DateTimeField):
+            if settings.USE_TZ:
+                if value is None:
+                    raise ValueError(
+                        # TODO (Josh): Change Error Message
+                        "Database returned an invalid value in QuerySet.datetimes(). "
+                        "Are time zone definitions for your database and pytz installed?"
+                    )
+                value = value.replace(tzinfo=None)
+                value = timezone.make_aware(value, self.tzinfo)
+        elif isinstance(self.output_field, DateField) and isinstance(value, datetime):
+            value = value.date()
+        return value
+
+
+class Trunc(TruncBase):
+
+    def __init__(self, expression, kind, output_field=None, tzinfo=None, **extra):
+        self.kind = kind
+        self.tzinfo = tzinfo
+        if not isinstance(output_field, (DateField, DateTimeField)):
+            raise ValueError('output_field must be either DateField or DateTimeField')
+        super(Trunc, self).__init__(expression, output_field=output_field, **extra)
+
+
+class TruncDateField(TruncBase):
+
+    @cached_property
+    def output_field(self):
+        return DateField()
+
+    def convert_value(self, value, expression, connection, context):
+        if isinstance(value, datetime):
+            value = value.date()
+        return value
+
+
+class TruncDateTimeField(TruncBase):
+
+    def __init__(self, expression, tzinfo=None, **extra):
+        super(TruncDateTimeField, self).__init__(expression, tzinfo=tzinfo, **extra)
+
+    @cached_property
+    def output_field(self):
+        return DateTimeField()
+
+
+class TruncYear(TruncDateField):
+    kind = 'year'
+
+
+class TruncMonth(TruncDateField):
+    kind = 'month'
+
+
+class TruncDay(TruncDateField):
+    lookup_name = 'date'
```

Comment:
NVM since it's both used as the `__date` lookup and as but truncation API the name sounds good.